napari/plugins/_npe2.py

```diff
@@ -233,9 +233,16 @@ def iter_manifests(
 
 def widget_iterator() -> Iterator[Tuple[str, Tuple[str, Sequence[str]]]]:
     # eg ('dock', ('my_plugin', ('My widget', MyWidget)))
+    # Make a dict of display_name matching plugin_name
+    plugin_names = {}
+    for plugin in pm.iter_manifests():
+        plugin_names[plugin.name] = plugin.display_name
+
     wdgs: DefaultDict[str, List[str]] = DefaultDict(list)
     for wdg_contrib in pm.iter_widgets():
-        wdgs[wdg_contrib.plugin_name].append(wdg_contrib.display_name)
+        wdgs[plugin_names[wdg_contrib.plugin_name]].append(
+            wdg_contrib.display_name
+        )
```

Comment:
I think something like this will work for the body of the for loop:

```suggestion
        # The plugin_display_name is distinct from the wdg_contrib.display_name
        # One plugin might have many widgets.
        plugin_display_name=pm.get_manifest(wdg_contrib.plugin_name)
        wdgs[plugin_display_name].append(
            wdg_contrib.display_name
        )
```

It avoids having to build the `plugin_names` dict.

I added the comment bc I get very confused about display_names :)

Reply:
I tested your code and it didn't work, so I didn't commit the suggestion
```
TypeError: arguments did not match any overloaded call:
  QMenu(parent: QWidget = None): argument 1 has unexpected type 'PluginManifest'
  QMenu(str, parent: QWidget = None): argument 1 has unexpected type 'PluginManifest'
```
I fixed it with:
`plugin_display_name=pm.get_manifest(wdg_contrib.plugin_name).display_name`